eth/downloader/whitelist/service.go

```diff
@@ -70,6 +72,96 @@ func (w *Service) IsValidChain(remoteHeader *types.Header, fetchHeadersByNumber
 	return false, ErrCheckpointMismatch
 }
 
+// IsValidChain checks the validity of chain by comparing it
+// against the local checkpoint entries
+func (w *Service) IsValidChain(currentHeader *types.Header, chain []*types.Header) bool {
+	var (
+		oldestCheckpointNumber uint64 = w.checkpointOrder[0]
+		first                  uint64 = chain[0].Number.Uint64()
+		last                   uint64 = chain[len(chain)-1].Number.Uint64()
+		current                uint64 = currentHeader.Number.Uint64()
+	)
+
+	// Check if we have checkpoints to validate incoming chain in memory
+	if len(w.checkpointWhitelist) == 0 {
+		// We don't have any entries, no additional validation will be possible
+		return true
```

Comment:
won't it be possible to feed new peers with malicious chains?

Reply:
So, do we want to consider the chain invalid, if we don't have local whitelist? As we just have a background process to gather whitelist and we only store them in memory I kind of think that there's good possibility of this map being empty or having old data. Wdyt?

Reply:
ask from Heimdall for N checkpoints at the start of the service.

Reply:
Done!